test/cbmc/sources/core_json_contracts.c

```diff
@@ -0,0 +1,185 @@
+
+JSONStatus_t JSON_ValidatePreconditions( char * buf,
+                                         size_t max )
+{
+    return ( max < CBMC_MAX_BUFSIZE )
+           & ( buf == NULL || allocated( buf, max ) );
+}
+
+bool JSON_IteratePostconditions( JSONStatus_t result,
+                                 char * buf,
+                                 size_t max,
+                                 JSONPair_t * outPair )
+{
+    CHECK( isJSONIterateEnum( result ) );
+
+    if( result == JSONSuccess )
+    {
+/* *INDENT-OFF* */
+        CHECK( ( outPair->key != NULL ) ==> ( ( outPair->key > buf ) && ( ( outPair->key + outPair->keyLength ) < ( buf + max ) ) ) );
+        CHECK( ( outPair->key != NULL ) ==> ( ( outPair->key + outPair->keyLength ) < outPair->value ) );
+/* *INDENT-ON* */
+        CHECK( ( outPair->value > buf ) && ( ( outPair->value + outPair->valueLength ) <= ( buf + max ) ) );
+        CHECK( isJSONTypesEnum( outPair->jsonType ) );
+    }
+
+    return true;
+}
+
+bool arraySearchPreconditions( char * buf,
+                               size_t max,
+                               size_t * outValue,
+                               size_t * outValueLength )
+{
+    return ( isValidBoundedBuffer( buf, max ) )
+           & ( allocated( outValue, sizeof( *outValue ) ) )
+           & ( allocated( outValueLength, sizeof( *outValueLength ) ) )
+           & ( *outValueLength <= max );
+}
+
+bool objectSearchPreconditions( char * buf,
+                                size_t max,
+                                const char * query,
+                                size_t queryLength,
+                                size_t * outValue,
+                                size_t * outValueLength )
+{
+    return arraySearchPreconditions( buf, max, outValue, outValueLength )
+           & ( queryLength < CBMC_MAX_QUERYKEYLENGTH )
+           & ( allocated( query, queryLength ) );
+}
+
+bool arraySearchPostconditions( bool result,
+                                char * buf,
+                                size_t max,
+                                size_t * outValue,
+                                size_t * outValueLength,
+                                size_t old_outValue,
+                                size_t old_outValueLength )
+{
+    if( result )
+    {
+        CHECK( 0 <= *outValue && *outValue < max );
+        CHECK( 0 < *outValueLength && *outValueLength <= max - *outValue );
+/* *INDENT-OFF* */
+        CHECK( buf[ *outValue ] == '"' ==> ( 2 <= *outValueLength && *outValueLength <= max - *outValue ) );
+/* *INDENT-ON* */
+    }
+    else
+    {
+        CHECK( *outValue == old_outValue );
+        CHECK( *outValueLength == old_outValueLength );
+    }
+
+    return true;
+}
+
+bool skipCollectionPostconditions( JSONStatus_t result,
+                                   char * buf,
+                                   size_t * start,
+                                   size_t old_start,
+                                   size_t max )
```

Comment:
as mentioned in the .h file, the skip contracts are spread out between macros,  functions and inline clauses, and  also between header and c files, it's hard to understand and maintain. I'd be in favor of using the same presentation for all contracts with a precondition function, a post condition function and an assigns clause function that have a name derived from the function

Reply:
Implemented as suggested.